### ISSUE TITLE
more resources

### ISSUE THREAD

**Issue by thorade** [ORNL-Modelica/awesome-modelica#3]
https://twitter.com/modelica
https://twitter.com/modelicainsider
https://twitter.com/openmodelica

https://www.modelon.com/news-blog

**Comment by greenwoodms06** [comment]
i added a couple of the twitters and modelon's blog. I'm leaning towards skipping openmodelica for reasons.... though probably bad ones. 

As a general point, I didn't initially intend for this to be an all-encompassing modelica resource page. more of a curated one of things that are the MOST important. I want to avoid someone coming along and just getting lost in a hyperlink forest. Thoughts? Do you find this useful?

**Comment by thorade** [comment]
I think this is useful, and I think several people have started such lists in various places, but it gets easily outdated, so maybe github is a good place to have this, so people can contribute and update it. Please skip whatever suggestion seems less relevant in your opinion. Also see #5 for similar lists in other areas

**Comment by thorade** [comment]
Ah, and regarding OpenModelica or other tools: I learned that it is possible to write libraries that work in several tools, e.g. most Modelon libraries work in Dymola and also in Modelon Impact (disclaimer: I work at Modelon). 
Step 1 is usually to turn on Dymola's pedantic model

**Comment by greenwoodms06** [comment]
i am going to take your suggestion and change the name to awesome-modelica. ill expand things a bit too, even open modelica :) rising tide lifts all boats and all that

**Comment by greenwoodms06** [comment]
added resources

**Comment by greenwoodms06** [comment]
@thorade question for ya.... I am in the process of trying to generate some general community material for getting started in modelica... as part of that, it crossed my mind that some sort of style guide should be created or explained. I know you asked about this a while back on [StackOverflow](https://stackoverflow.com/questions/14163386/modelica-style-guide). Did you ever get anything going? If not I may try to cobble something together over time.

**Comment by greenwoodms06** [comment]
i know that it could be highly dependent on area (see Tiller's comment) but I was thinking really just general practice things like PascalCase for classes, camelCase for instances of the classes and function names, recommended package structures (package, examples, models, components, etc...), use of import for common things like `import SI = Modelica.SI.Units`, ways to annotate, label, code commenting (or not), documentation page, etc.

**Comment by thorade** [comment]
What I was really hoping for when asking that, was the second part of the question: 
And, ideally, a tool that helps with the re-formatting, indentation etc.?

Because I was so annoyed by the many unnecessary whitespace changes that Dymola often introduces. There are some other stackoverflow questions also about Dymola whitespace auto-formatting, especially with version control. These formatting changes make it much more timeconsuming to do reviews and they will cause merge conflicts.
It would be great to 

- stop Dymola from doing that auto-formatting, 
- and as a second step maybe have rules like a subset of PEP8, for example: recommend wide equal, space after comma, function inputs on same line or not, trailing whitespace, 
- and as a third step have a tool like autopep8 or black for python to reformat the code.

Regarding conventions for variable names, I think there will never be something universal. Most important is maybe to be consistent within one project. Then some common conventions like classnames start with uppercase, functions with lowercase.

The only additions to the SO question are maybe these two:
https://simulationresearch.lbl.gov/modelica/userGuide/development.html#style-guide
https://github.com/dietmarw/trimtrailingwhitespaces

**Comment by thorade** [comment]
Some Modelica-specific recommendation for beginners:
Avoid Real, always use types with min, max, nominal, start and unit.
Use the description string:
`parameter SI.Temperature T_min "minimal Temperature";` is much better than
`parameter Real T;`

Another basic recommendation: Dymola has the check button, use it early and often and address all the warnings and errors. For a library developed from scratch, start with pedantic mode right away!

And I still like the common sense recommendations in the Matlab pdf I linked in the SO question.

**Comment by greenwoodms06** [comment]
What!?!?! You don't like popping over to your version control and having it tell you you've changed 300 files magically when all you did was open Dymola up? 😆 Yeah... I think it has at least gotten better in the newer versions but still happens.

It seems to me that Dymola at least could add a clean up feature pretty easily. Oh well. Add to my list of things to implement someday.

All good points... This has me thinking of making a `style-guide-modelica` repo. Do you know of a style-guide-guide like the `awesome` repo you pointed me to? There was this [style-guide-guide](https://github.com/bradfrost/style-guide-guide) but it seemed a bit much for what we are discussing. A simple markdown document would seem sufficient.

I could start codifying these types of things for projects more generally. Basically it would be something people could point to and then adapt as they need for their project. This might very well be something that then could morph into some sort of standard pep8 type thang. I think I would try to mush the information from LBNL's guide and the Matlab guide's approach into a markdown document. Erring on the side of less complex/complicated...

**Comment by Satinelamp** [comment]
Maybe a good initialization strategy would be useful. For example, the library [ThermoSysPro](https://github.com/ThermoSysPro/ThermoSysPro) from EDF is NOT using any initialization parameters and give the iteration variables the initialization guess values directly, so it could use a script to update the iteration variables to make the initialization convergence process robust.
![Snipaste_2021-03-07_09-40-11](https://user-images.githubusercontent.com/62006565/110226405-d0e53100-7f29-11eb-8628-43f6dc1a2ca3.jpg)

The main idea is explained in the following chapter of [Modeling and Simulation of Thermal Power Plants with ThermoSysPro](https://www.springer.com/gp/book/9783030051044)
![G7710](https://user-images.githubusercontent.com/62006565/110226352-66cc8c00-7f29-11eb-9988-cd83b6e07ba9.png)
![3xBim](https://user-images.githubusercontent.com/62006565/110226360-764bd500-7f29-11eb-82b3-3f8e94af510b.png)

But there could also some other strategies, as used in the MSL, it use some initialization parameters to make it more user-friendly to change the initialization settings, but it might be hard to [avoid transient behavior at startup](https://www.claytex.com/blog/methods/avoiding-transient-behaviour-at-startup-storing-start-values-in-a-model/)
And when initialization failure occurs, it is hard to adjust these initialization guess values by hand.
![Snipaste_2021-03-07_09-44-47](https://user-images.githubusercontent.com/62006565/110226439-43561100-7f2a-11eb-8e6b-5fbf04142548.jpg)

To overcome the issue of transient behavior at startup, some other library uses a totally static model which need to build a companion sublibrary to provide static components, and it uses the results from the static cycle to provide initialization guess values to make the convergence robust.

![Snipaste_2021-03-07_09-54-26](https://user-images.githubusercontent.com/62006565/110226521-2241f000-7f2b-11eb-9f25-0094c725628e.jpg)

To summarise all mentioned above, I think for large-scale systems, there should be a discussion, maybe even a standard about how to do initialization and make it for users easier to understand and control the initialization process.

**Comment by greenwoodms06** [comment]
@Satinelamp Indeed the essence of what you are saying (i.e., the many difficulties with Modelica) I would boil to down one word: "art". IMO there is too much art involved into creating even mildly complex modelica models. For modelica to grow, it needs:
1. To get standardized workflows/guidance for modeling areas (e.g, thermal hydraulics). In this I include not only the types of things you are mentioning with initialization, but also the best practice for creating components (e.g., when and how best to initialize ports, etc.).
2. More robust and functional open-source tools. For example, each step in the Modelica to c code process could probably be cut up and then sent to individual organizations who can focus their efforts on that one aspect. Then even these piece parts can be swapped out and improved to move towards a really paradigm shifting capability.
3. Did I mention robust and open-source (i.e., free) tools? Pay to play is not a viable strategy for Modelica and unfortunately that is where this language is stuck at IMO. This has led to concentration of know-how to commercial companies/developers and has, I think, significantly restricted the reach of Modelica. Julia has exploded over a short period of time. Who knows where it will go but why shouldn't have Modelica done the same!?

But to get back to your point. There is a lot that could be done to reduce the amount of expert knowledge (i.e., art) to create complex and useful models in Modelica. I think a complete redesign and expansion of the MSL fluids library would be a natural output of that effort. In my mind this would start with a working group focused on thermal-hydraulic modeling. That working group would develop the standard approach for this work, discover/describe new methods to improve simulation, advise vendors/developers on new capabilities that should add, etc., etc. To me it is important to stress the idea that this group would not delve as little as possible into specific areas of TH (e.g., power systems or buildings) but would finally articulate in document and models how things should be done. A new and expanded version of the MSL fluids library would be good. Or frankly a competitor (a la IBPSA) to MSL fluids. From there more tailored libraries and workflows could be created to industries (e.g., power plants).

**Comment by greenwoodms06** [comment]
I also have had thoughts about actually a new Modelica-like language. To me, I have always kind of had issues with "parameters" vs "inputs" vs "Real-only" vs "outputs". In reality each variable should be a mini class with various flags, start values, etc. that are accessible from the editor (see code below). The work is then shifted entirely to the translator recognize how the equations/data are used. This way the user can make anything (in theory) time dependent. Start values should be able to be robustly set. etc. Anyways, I can't say this portion of my rant is necessarily the best way to go but it's a discussion piece. 

````
Real x;
Real y(start=x.start);
````
